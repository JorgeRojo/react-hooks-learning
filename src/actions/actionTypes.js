// COUNTER
export const COUNTER_DECREASE = 'COUNTER_DECREASE';
export const COUNTER_INCREASE = 'COUNTER_INCREASE';
export const COUNTER_RESET = 'COUNTER_RESET';
export const COUNTER_SET = 'COUNTER_SET';
// HTTP_REQUEST
export const HTTP_REQUEST = 'HTTP_REQUEST';
// API
export const API = 'API';
export const API_REQUEST = 'API/REQUEST';
export const API_ERROR = 'API/ERROR';
export const API_SUCCESS = 'API/SUCCESS';
export const API_RESET = 'API/RESET';
