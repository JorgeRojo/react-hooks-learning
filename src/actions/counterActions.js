import {
  COUNTER_DECREASE,
  COUNTER_INCREASE,
  COUNTER_RESET,
  COUNTER_SET,
} from './actionTypes';

export const counterDecrease = () => ({
  type: COUNTER_DECREASE,
});

export const counterIncrease = () => ({
  type: COUNTER_INCREASE,
});

export const counterReset = () => ({
  type: COUNTER_RESET,
});

export const counterSet = number => ({
  type: COUNTER_SET,
  payload: { number },
});
