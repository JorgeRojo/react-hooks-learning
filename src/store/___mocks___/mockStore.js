import configureMockStore from "redux-mock-store";
import thunk from "redux-thunk";
import mockState from "./mockState";
import withReduxDispatchGenerator from "../withReduxDispatchGenerator";

const middleware = [thunk];
const mockStore = configureMockStore(middleware);

const store = mockStore(mockState);

export default store;

export const withReduxDispatch = withReduxDispatchGenerator(store);
