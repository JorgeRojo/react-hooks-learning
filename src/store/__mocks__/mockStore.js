import configureMockStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import mockState from './mockState';

const middleware = [thunk];
const mockStore = configureMockStore(middleware);

const store = mockStore(mockState);

export default store;
