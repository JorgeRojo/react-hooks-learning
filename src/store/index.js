import { createStore, applyMiddleware } from 'redux';
import { composeWithDevTools } from 'redux-devtools-extension';
import thunk from 'redux-thunk';
import reducer from '../reducer';

const initialState = {};
const middleware = [thunk];
const composeEnhancers = composeWithDevTools({});

const store = createStore(
  reducer,
  initialState,
  composeEnhancers(applyMiddleware(...middleware))
);

export default store;
