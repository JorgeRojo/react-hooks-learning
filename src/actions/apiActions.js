import { HTTP_REQUEST, API_RESET, API } from './actionTypes';
import { HttpRequest } from '../store/httpRequestReduxMiddleware';

export const apiGet = () => ({
  type: HTTP_REQUEST,
  nextActionType: API,
  request: new HttpRequest('https://jsonplaceholder.typicode.com/posts'),
  responseStatus: 200,
});

export const apiReset = () => ({
  type: API_RESET,
});
