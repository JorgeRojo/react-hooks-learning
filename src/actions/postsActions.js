import { HTTP_REQUEST, POSTS_RESET, POSTS } from './actionTypes';
import { HttpRequest } from '../store/httpRequestReduxMiddleware';

export const postsGet = () => ({
  type: HTTP_REQUEST,
  nextActionType: POSTS,
  request: new HttpRequest('https://jsonplaceholder.typicode.com/posts'),
  responseStatus: 200,
});

export const postsReset = () => ({
  type: POSTS_RESET,
});
