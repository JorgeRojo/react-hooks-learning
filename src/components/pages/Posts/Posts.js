import React, { Component } from 'react';
import { connect } from 'react-redux';
import { selectPostsList, selectPostsIsLoading } from '../../../reducer';
import { postsGet, postsReset } from '../../../actions/postsActions';

class Posts extends Component {
  componentDidMount() {
    this.props.postsGet();
  }

  componentWillUnmount() {
    this.props.postsReset();
  }

  render() {
    const { isLoading, posts } = this.props;
    return (
      <div className="container text-center mt-4">
        <h1 className="display-4 mb-4">Posts list async response</h1>
        {isLoading && <p>Is Loading...</p>}
        {!isLoading && posts.length > 0 && (
          <>
            {posts.map(({ title, body }, index) => (
              <div className="card mb-3" key={index}>
                <div className="card-header">{title}</div>
                <div className="card-body">{body}</div>
              </div>
            ))}
          </>
        )}
      </div>
    );
  }
}

const mapStateToProps = state => ({
  isLoading: selectPostsIsLoading(state),
  posts: selectPostsList(state),
});

const mapDispatchToProps = {
  postsGet,
  postsReset,
};

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(Posts);
