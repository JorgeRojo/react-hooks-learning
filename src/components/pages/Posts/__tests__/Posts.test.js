import React from 'react';
import ReactDOM from 'react-dom';
import { render, cleanup } from '@testing-library/react';
import { Provider } from 'react-redux';

import store from '../../../../store';
import { postsGet, postsReset } from '../../../../actions/postsActions';
import {
  selectPostsList,
  selectPostsIsLoading,
} from '../../../../reducer/posts';

import Posts from '../Posts';
import mockState from '../../../../store/__mocks__/mockState';

jest.mock('../../../../store', () =>
  require('../../../../store/__mocks__/mockStore')
);
jest.mock('../../../../actions/postsActions');
jest.mock('../../../../reducer/posts');

describe('Home Component', () => {
  const component = (
    <Provider store={store}>
      <Posts />
    </Provider>
  );

  beforeEach(() => {
    selectPostsIsLoading.mockReturnValue(false);
    selectPostsList.mockReturnValue(mockState.posts.response.data);
    postsGet.mockImplementation(() => ({ type: 'postsGet' }));
    postsReset.mockImplementation(() => ({ type: 'postsReset' }));
    postsReset.mockImplementation(() => ({ type: 'postsReset' }));
  });

  afterEach(() => {
    cleanup();
    jest.clearAllMocks();
  });

  it('renders without crashing', () => {
    const div = document.createElement('div');
    ReactDOM.render(component, div);
    ReactDOM.unmountComponentAtNode(div);
  });

  it('renders the component', () => {
    const { container } = render(component);
    expect(container).toMatchSnapshot();
  });

  it('renders the component with is loading', () => {
    selectPostsIsLoading.mockReturnValue(true);
    const { container } = render(component);
    expect(container).toMatchSnapshot();
  });
});
