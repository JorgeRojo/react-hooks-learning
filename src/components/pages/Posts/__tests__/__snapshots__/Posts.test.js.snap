// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home Component renders the component 1`] = `
<div>
  <div
    class="container text-center mt-4"
  >
    <h1
      class="display-4 mb-4"
    >
      Posts list async response
    </h1>
    <div
      class="card mb-3"
    >
      <div
        class="card-header"
      >
        sunt aut facere repellat provident occaecati excepturi optio reprehenderit
      </div>
      <div
        class="card-body"
      >
        quia et suscipit
suscipit recusandae consequuntur expedita et cum
reprehenderit molestiae ut ut quas totam
nostrum rerum est autem sunt rem eveniet architecto
      </div>
    </div>
    <div
      class="card mb-3"
    >
      <div
        class="card-header"
      >
        qui est esse
      </div>
      <div
        class="card-body"
      >
        est rerum tempore vitae
sequi sint nihil reprehenderit dolor beatae ea dolores neque
fugiat blanditiis voluptate porro vel nihil molestiae ut reiciendis
qui aperiam non debitis possimus qui neque nisi nulla
      </div>
    </div>
    <div
      class="card mb-3"
    >
      <div
        class="card-header"
      >
        ea molestias quasi exercitationem repellat qui ipsa sit aut
      </div>
      <div
        class="card-body"
      >
        et iusto sed quo iure
voluptatem occaecati omnis eligendi aut ad
voluptatem doloribus vel accusantium quis pariatur
molestiae porro eius odio et labore et velit aut
      </div>
    </div>
    <div
      class="card mb-3"
    >
      <div
        class="card-header"
      >
        eum et est occaecati
      </div>
      <div
        class="card-body"
      >
        ullam et saepe reiciendis voluptatem adipisci
sit amet autem assumenda provident rerum culpa
quis hic commodi nesciunt rem tenetur doloremque ipsam iure
quis sunt voluptatem rerum illo velit
      </div>
    </div>
    <div
      class="card mb-3"
    >
      <div
        class="card-header"
      >
        nesciunt quas odio
      </div>
      <div
        class="card-body"
      >
        repudiandae veniam quaerat sunt sed
alias aut fugiat sit autem sed est
voluptatem omnis possimus esse voluptatibus quis
est aut tenetur dolor neque
      </div>
    </div>
  </div>
</div>
`;

exports[`Home Component renders the component with is loading 1`] = `
<div>
  <div
    class="container text-center mt-4"
  >
    <h1
      class="display-4 mb-4"
    >
      Posts list async response
    </h1>
    <p>
      Is Loading...
    </p>
  </div>
</div>
`;
